libavcodec/libaribcaption.c

```diff
@@ -0,0 +1,1141 @@
+    int outline, shadow;
+    const char *font_name;
+    const char *fonts = ctx->font;
+
+    if (ctx->border_style == 4) {
+        outline = 0;
+        shadow = 4;
+    } else {
+        outline = 1;
+        shadow = 0;
+    }
+    if (ctx->force_stroke_text)
+        outline = (int)(ctx->stroke_width * 4.0 / 3.0);
+
+    if (fonts && *fonts)
+        font_name = av_get_token(&fonts, ",");
+    else
+        font_name = av_strdup(DEFAULT_FONT_FAMILY);
+    if (!font_name)
+        return AVERROR(ENOMEM);
+
+    if (avctx->subtitle_header)
+        av_freep(&avctx->subtitle_header);
+    avctx->subtitle_header = av_asprintf(
+            "[Script Info]\r\n"
+            "ScriptType: v4.00+\r\n"
+            "PlayResX: %d\r\n"
+            "PlayResY: %d\r\n"
+            "WrapStyle: 2\r\n"      /* 2: no word wrapping */
+            "\r\n"
+
+            "[V4+ Styles]\r\n"
+             "Format: Name, "
+             "Fontname, Fontsize, "
+             "PrimaryColour, SecondaryColour, OutlineColour, BackColour, "
+             "Bold, Italic, Underline, StrikeOut, "
+             "ScaleX, ScaleY, "
+             "Spacing, Angle, "
+             "BorderStyle, Outline, Shadow, "
+             "Alignment, MarginL, MarginR, MarginV, "
+             "Encoding\r\n"
+
+             "Style: "
+             "Default,"             /* Name */
+             "%s,%d,"               /* Font{name,size} */
+             "&H%x,&H%x,&H%x,&H%x," /* {Primary,Secondary,Outline,Back}Colour */
+             "%d,%d,%d,0,"          /* Bold, Italic, Underline, StrikeOut */
+             "100,100,"             /* Scale{X,Y} */
+             "0,0,"                 /* Spacing, Angle */
+             "%d,%d,%d,"            /* BorderStyle, Outline, Shadow */
+             "%d,10,10,10,"         /* Alignment, Margin[LRV] */
+             "0\r\n"                /* Encoding */
+             "\r\n"
+
+             "[Events]\r\n"
+             "Format: Layer, Start, End, Style, Name, MarginL, MarginR, MarginV, Effect, Text\r\n",
+            ctx->plane_width, ctx->plane_height,
+            font_name, ctx->font_size,
+            ASS_DEFAULT_COLOR, ASS_DEFAULT_COLOR,
+            ASS_DEFAULT_BACK_COLOR, ASS_DEFAULT_BACK_COLOR,
+            -ASS_DEFAULT_BOLD, -ASS_DEFAULT_ITALIC, -ASS_DEFAULT_UNDERLINE,
+            ctx->border_style, outline, shadow, ASS_DEFAULT_ALIGNMENT);
+
+    av_freep(&font_name);
+    if (!avctx->subtitle_header)
+        return AVERROR(ENOMEM);
+    avctx->subtitle_header_size = strlen(avctx->subtitle_header);
+    return 0;
+}
+
+static void set_ass_color(AVBPrint *buf, int color_num,
+                          aribcc_color_t new_color, aribcc_color_t old_color)
+{
+    if (ARIBCC_COLOR_DIFF_RGB(new_color, old_color))
+        av_bprintf(buf, "{\\%dc&H%06x&}", color_num,
+                                          ARIBCC_COLOR_RGB(new_color));
+    if (ARIBCC_COLOR_DIFF_A(new_color, old_color) && ARIBCC_COLOR_A(new_color))
```

Comment:
This should check if the alpha is _different_; if it changes to 0, we should set that.

Reply:
Actual ISDB broadcasting doesn't allways set alpha value, and 0 is set in many cases.
Alpha value 0 means transparent, and as the result subtitle cannot seen if we strictly obey it.

Reply:
0 in ASS means fully opaque; _0xFF_ means fully transparent. I think you need to invert the value here.

I don't see any corresponding code in libaribcaption's renderer to ignore 0 values?

Reply:
Is it true?
If so, above line 559 should not changed. Instead, ASS style override code number would be `0xFF - alpha`.

Reply:
Correct, though I'm still confused about why this check is necessary when libaribcaption's renderer doesn't seem to have anything equivalent (unless I'm missing something?). In any case, it at least needs a comment explaining why.

Reply:
I checked some TS files on my hand. I can't find text color with alpha 0. So, I might see apparition.
I'll remove the check.